contributors/design-proposals/raw-block-pv.md

```diff
@@ -0,0 +1,571 @@
+spec:
+  serviceName: "nginx"
+  replicas: 2
+  template:
+    metadata:
+      labels:
+	app: nginx
+    spec:
+      containers:
+      - name: nginx
+	image: gcr.io/google_containers/nginx-slim:0.8
+	ports:
+	- containerPort: 80
+	  name: web
+	volumeMounts:
+	- name: www
+	  mountPath: /usr/share/nginx/html
+ volumeClaimTemplates:
+  - metadata:
+      name: datadir
+      annotations:
+	volume.beta.kubernetes.io/storage-class: aws-ebs-raw
+    spec:
+      accessModes:
+	- "ReadWriteOnce"
+      resources:
+	requests:
+	  storage: 10Gi
+```
+***SUITABLE FOR: NETWORK ATTACHED BLOCK***
+
+## UC8: 
+
+DESCRIPTION: Admin creates network raw block devices
+
+BACKGROUND: Admin wishes to pre-create Persistent Volumes in GCE as raw block devices
+
+WORKFLOW:
+
+ADMIN:
+
+```
+apiVersion: "v1"
+kind: "PersistentVolume"
+metadata:
+  name: gce-disk-1
+  annotations:
+    volume.beta.kubernetes.io/mount-options: "discard"
+Spec:
+  volumeType: block
+  capacity:
+    storage: "10Gi"
+  accessModes:
+    - "ReadWriteOnce"
+  gcePersistentDisk:
+    fsType: "raw"
+    pdName: "gce-disk-1"
+
+```
+
+***If admin specifies volumeType: block + fstype: ext4 then they would get what they already get today ***
+
+## UC9: 
+
+DESCRIPTION: Developer wishes to consumes raw device
+
+WORKFLOW:
+
+USER:
+
+```
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+  name: pvc-raw-block
+spec:
+  volumeType: block
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 80Gi
+```
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: my-db
+spec:
+    containers:
+    - name: mysql
+      image: mysql
+      volumeMounts:
+      - name: my-db-data
+	mountPath: /var/lib/mysql/data
+    volumes:
+    - name: my-db-data
+      persistentVolumeClaim:
+	claimName: pvc-raw-block
+```
+
+# Implementation Plan
+
+Phase 1: Pre-provisioned PVs to precreated devices
+
+               API changes
+
+               Special considerations: Access, Scrubbing device after PV is deleted
+               
+               Changes to the mounter interface as today it is assumed 'file' as the default.
+
+Phase 2:  Discovery of block devices 
+
+                Dynamically provisioned PVs to dynamically allocated devices
+
+                Priv container concerns    
+
+Other considerations:
+
+               Expose volumeType to users via kubectl
+
+               Adds enable/disable configuration to securityContext
+
+               Reference volume driver change(Attach/Detach logic) for pre-provisioned PVs
+
+               Reference volume driver change(Provision/Delete logic) for dynamic provisioning
+
+## Mounter interface proposed design
+
+```
+ type BlockMounter interface {
+	Mounter
+ 	CanSupportFileOnBlock() error
+            // kubelet needs volume's device path when attaches a volume to container
+ 	GetVolumePath() string
+ 	GetVolumeType() string   //TBD
+ }
+
+type BlockUnmounter interface {
+ 	Unmounter
+ 	GetVolumePath() string
+}
+```
```

Comment:
Please add "Features & Milestones" section, then we need to mention that there are two phases to support of block volume support.